x-pack/plugins/dataset_quality/public/state_machines/dataset_quality_controller/src/state_machine.ts

```diff
@@ -174,6 +175,24 @@ export const createPureDatasetQualityControllerStateMachine = (
               }
             : {};
         }),
+        storeDatasets: assign((context, _event) => {
+          return context.dataStreamStats && context.degradedDocStats
+            ? {
+                datasets: context.dataStreamStats?.map((dataStream) => {
+                  const degradedDocs = find(context.degradedDocStats, {
+                    dataset: dataStream.rawName,
+                  });
+
+                  return {
+                    ...dataStream,
+                    degradedDocs: degradedDocs?.percentage,
+                  };
+                }),
```

Comment:
**suggestion**: this works well on low-mid range data amounts, but as we don't have insights on how many datasets can users have, this could have an impact on the UI performance being a blocking task with quadratic complexity.

For instance, I tried to run it on a hypothetical account with 1000+ datasets (with 1000+ respective degraded docs entries), which resulted on the following result:

<img width="389" alt="Screenshot 2024-01-31 at 09 34 46" src="https://github.com/elastic/kibana/assets/34506779/81563577-e57c-462f-97ac-4dd6b1211d7c">

Even on high end machines it takes a long time, I suggest a quick refactor to make this operation complexity linear using a map:
```ts
function mergeDegradedPercentageIntoDataStreams(dataStreamStats, degradedDocStats) {
  const degradedMap = degradedDocStats.reduce(
    (degradedMapAcc, { dataset, percentage }) =>
      Object.assign(degradedMapAcc, { [dataset]: percentage }),
    {}
  );

  return dataStreamStats?.map((dataStream) => ({
    ...dataStream,
    degradedDocs: degradedMap[dataStream.name],
  }));
}
```

Then you could use it in your action :)